Installation Folder & rpath urls

gabyx opened the pull request:
the same as in #101 ,

gabyx commented:
![bildschirmfoto 2018-01-21 um 22 29 16](https://user-images.githubusercontent.com/647437/35199170-c870e026-fefa-11e7-979e-f33acf15e72a.png)

This is now pretty fine -> executable work in build `bin` folder (on macos)
After installation the binary no longer work, because the rpaths have not been adapted. Cmake should actually do something when installing (see https://cmake.org/Wiki/CMake_RPATH_handling)

the output of `otool -l /usr/local/opt/rttr/bin/bench_variant | grep LC_RPATH -A2` 
```
          cmd LC_RPATH
      cmdsize 40
         path /usr/local/opt/llvm/lib (offset 12)
```
where as the binary in the build folder works and has
```
         cmd LC_RPATH
      cmdsize 40
         path /usr/local/opt/llvm/lib (offset 12)
--
          cmd LC_RPATH
      cmdsize 56
         path /Users/gabrielnuetzi/Desktop/rttrBuild/lib (offset 12)
```
this is the reason it work in the build folder, I dont understand why cmake has not correctly adapted the RPATH after installing, there is a misconfiguration I think...

acki-m commented on CMake/utility.cmake:
```diff
@@ -165,7 +165,7 @@ function(loadFolder FOLDER _HEADER_FILES _SOURCE_FILES)
     getNameOfDir(CMAKE_CURRENT_SOURCE_DIR DIRNAME)
     if (${shouldInstall})
       if (NOT ${FULL_HEADER_PATH} MATCHES ".*_p.h$") # we don't want to install header files which are marked as private
-        install(FILES ${FULL_HEADER_PATH} DESTINATION "include/${DIRNAME}/${REL_PATH}" PERMISSIONS OWNER_READ)
```
why is this removed? I want that the installed header files should not be modified by accident.

gabyx replied:
wrong merge, I will correct that

gabyx commented:
thanks, its a mistake!

Warchant commented on CMakeLists.txt:
```diff
@@ -31,7 +31,7 @@
 ####################################################################################
 
-cmake_minimum_required (VERSION 2.8.12)
+cmake_minimum_required (VERSION 3.10)
```
3.10 version is too strict, IMO 3.0 is enough, unless you use something specific.

gabyx replied:
thanks!
ok, cmake version 3.0 is fine for me, I rather work with the newest, since there is better behavior :-)

gabyx replied:
and it is not a problem for the user to have higher version

Warchant commented:
```bash
$ mkdir build
$ cd build
$ cmake .. 
$ make
```
fails with
```
[ 99%] Linking CXX executable ../../bin/unit_tests
[ 99%] Built target unit_tests
Scanning dependencies of target run_tests
[100%] Running unit_tests
../../bin/unit_tests: error while loading shared libraries: librttr_core.so.0.9.6: cannot open shared object file: No such file or directory
src/unit_tests/CMakeFiles/run_tests.dir/build.make:57: recipe for target 'src/unit_tests/CMakeFiles/run_tests' failed
make[2]: *** [src/unit_tests/CMakeFiles/run_tests] Error 127
CMakeFiles/Makefile2:206: recipe for target 'src/unit_tests/CMakeFiles/run_tests.dir/all' failed
make[1]: *** [src/unit_tests/CMakeFiles/run_tests.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

gabyx commented:
@Warchant : Could you extract some information in the unit_test binaries (I could not build them `catch` is too old for gcc 7 or clang 7).

I mean what is the RPATH the unit tests are search for the library, 

https://cmake.org/Wiki/CMake_RPATH_handling
` objdump -x executable_or_lib.so | grep RPATH`

I think we miss some configuration for `INSTALL_RPATH` here: https://github.com/rttrorg/rttr/blob/47d576c6b02ece18757fe0a07dd041e3a54280c6/src/unit_tests/CMakeLists.txt#L46

gabyx commented:
Building with gcc4.9 (macOs).... lets see whats the problem

gabyx commented:
ok found the error, I think I mistakenly misscorrected the beahvior for rpath...

gabyx commented:
lets look on the tests, the unit tests build with gcc4.9 on mac, libraries are loaded, but no executable does run (unit_tests and benchmarks),
: weird error: 
```
bench_variant(24239,0x7fff74ebc000) malloc: *** error for object 0x106ecbea0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
clock resolution: Abort trap: 6
```
@acki-m: Any idea? Supposed wrong linking, ...?
with clang7.0 and all warnings off I builded everythin except the unit_tests since catch is too old?
are you improving on this? meaning, installing a newer version!? that would be great!
So we can make some progress with some really more modern compilers instead of the ones supported, I think they are way behind...

gabyx commented:
seems to compile now, apple travis is missing, they have problems I think...

gabyx commented:
OK: only https://travis-ci.org/rttrorg/rttr/jobs/332456069 still fails due to some installation problems , probably re-run it again....

acki-m commented:
What are you fixing here? What is the problem on MacOSX?

gabyx commented:
Changed the PR title: I made more consistent cmake code, for linux and mac, windows, and set the settings for the rpath urls in the binaries.
as well as incorporated the changes of @Warchant (some of them)
Default prefix when configuring is still `./install`, these are some really valuable changes I think.

- [x]  Newer Cmake
- [x]  Made rpath settings in binaries consistent in Linux/Mac (Windows not affected since that thing does not exist in Win) (newer cmake has some new policy for macos, thats why it did not work)
- [x]  Better Install directories (default is `./install`)

These fixes are also incorporated for the PR #100 (which is on good way I think, since, clang 7.0 compiles fine with all warning on, except for benchmarks and unit_tests there I have disabled some minor ones... (see the PR)

gabyx commented:
Can we merge this?

acki-m commented:
I checked it out and compiled it with MSVC. It works so far, however.
I am unsure about:
* higher CMake version 3.5
* the switching of: `-stdlib=libstdc++ -static -lstdc++`
* `DESTINATION ${CMAKE_INSTALL_DATADIR}/rttr/cmake` What location is this?
* how to you copy the readme and the license file now?

PS: I would merge this first, because this fixes the path handling

gabyx commented:
that would be great!, 
- ${CMAKE_INSTALL_DATADIR}= /usr/local/share
on windows its set to the CMAKE_INSTALL_PREFIX
- higher CMake version 3.5 -> set it back to 3.0 in #100 
- readme file back added in #100

acki-m commented:
Damn't it I checked out master from your fork ^^
That's why everything worked. Have to do it again.

gabyx commented:
Oh pitty, ehm -> let me know if something is not working, so we can fix it =)

gabyx commented:
At least appveyor succeded

acki-m commented on CMake/config.cmake:
```diff
@@ -40,29 +40,56 @@ set(README_FILE "${CMAKE_SOURCE_DIR}/README.md")
+  set(CMAKE_INSTALL_DATADIR "${CMAKE_INSTALL_PREFIX}")
+endif()
+
+set(CMAKE_DEBUG_POSTFIX CACHE STRING "_d")
```
for unkown reason the postfix will not be applied to `rttr_core`
![image](https://user-images.githubusercontent.com/11368160/35354149-84c88d00-0149-11e8-8292-1118801222b5.png)

acki-m replied:
The postfix is necessary, otherwise we cannot copy the debug and release libraries, into one folder.

gabyx replied:
for sure! but why is it not working? the CMAKE_DEBUG_POSTFIX is set...?

acki-m commented on CMake/config.cmake:
```diff
@@ -134,10 +161,6 @@ write_basic_package_version_file(
+             DESTINATION ${CMAKE_INSTALL_DATADIR}/rttr/cmake
              COMPONENT Devel)
-
-    install(FILES "${LICENSE_FILE}" "${README_FILE}"
```
license files are missing

acki-m commented on CMake/config.cmake:
```diff
@@ -134,10 +161,6 @@ write_basic_package_version_file(
 if (BUILD_INSTALLER)
     install(FILES "${CMAKE_CURRENT_BINARY_DIR}/CMake/rttr-config-version.cmake"
-             DESTINATION cmake
+             DESTINATION ${CMAKE_INSTALL_DATADIR}/rttr/cmake
```
this looks now strange on windows:
![image](https://user-images.githubusercontent.com/11368160/35354238-be34cebe-0149-11e8-82d3-f92e445b7714.png)
The folder should be named cmake. I think otherwise cmake will not find the script, when pointing to the root directory

gabyx replied:
the lib folder should be empty? isnt it?  the libraries should all go into the bin folder...?

gabyx replied:
may be we could symlink on windows into the lib folder, and install all everythin properly in the lib folder?, why is the lib folder even constructed?

acki-m replied:
The lib folder has to be there, there are the *.lib files placed.
necessary to link the libraries on windows

gabyx replied:
ah, where is this path hardcoded into cmake? i didnt see that one? maybe I missed something?

acki-m commented on CMakeLists.txt:
```diff
@@ -31,7 +31,7 @@
 ####################################################################################
 
-cmake_minimum_required (VERSION 2.8.12)
+cmake_minimum_required (VERSION 3.5)
```
change to 3.0

gabyx commented:
fixed, strange with the debug post fix? does it work with a reconfigure?
> Am 24.01.2018 um 21:02 schrieb acki-m <notifications@github.com>:
> 
> @acki-m requested changes on this pull request.
> 
> In CMake/config.cmake <https://github.com/rttrorg/rttr/pull/103#discussion_r163661675>:
> 
> > -  set(RTTR_LIB_INSTALL_DIR "lib")
> -else()
> -  set(RTTR_BIN_INSTALL_DIR ${RTTR_TARGET_BIN_DIR})
> -  set(RTTR_LIB_INSTALL_DIR "lib")
> +
> +# set all install directories for the targets
> +if(UNIX)
> +  include(GNUInstallDirs)
> +  # all directories set
> +else(WINDOWS)
> +  set(CMAKE_INSTALL_LIBDIR "lib")
> +  set(CMAKE_INSTALL_BINDIR "bin")
> +  set(CMAKE_INSTALL_DATADIR "${CMAKE_INSTALL_PREFIX}")
> +endif()
> +
> +set(CMAKE_DEBUG_POSTFIX CACHE STRING "_d")
> for unkown reason the postfix will not be applied to rttr_core
>  <https://user-images.githubusercontent.com/11368160/35354149-84c88d00-0149-11e8-8292-1118801222b5.png>
> In CMake/config.cmake <https://github.com/rttrorg/rttr/pull/103#discussion_r163661872>:
> 
> >               COMPONENT Devel)
> -
> -    install(FILES "${LICENSE_FILE}" "${README_FILE}"
> license files are missing
> 
> In CMake/config.cmake <https://github.com/rttrorg/rttr/pull/103#discussion_r163662237>:
> 
> > @@ -134,10 +161,6 @@ write_basic_package_version_file(
>  
>  if (BUILD_INSTALLER)
>      install(FILES "${CMAKE_CURRENT_BINARY_DIR}/CMake/rttr-config-version.cmake"
> -             DESTINATION cmake
> +             DESTINATION ${CMAKE_INSTALL_DATADIR}/rttr/cmake
> this looks now strange on windows:
>  <https://user-images.githubusercontent.com/11368160/35354238-be34cebe-0149-11e8-82d3-f92e445b7714.png>
> The folder should be named cmake. I think otherwise cmake will not find the script, when pointing to the root directory
> 
> In CMakeLists.txt <https://github.com/rttrorg/rttr/pull/103#discussion_r163662288>:
> 
> > @@ -31,7 +31,7 @@
>  # the build configuration options are initialized.                                 #
>  ####################################################################################
>  
> -cmake_minimum_required (VERSION 2.8.12)
> +cmake_minimum_required (VERSION 3.5)
> change to 3.0
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/rttrorg/rttr/pull/103#pullrequestreview-91311194>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAnhDQG4qAjm79TqPoHTUAc2ljUBP_Pqks5tN4xugaJpZM4Rl8ID>.
> 

Gabriel Nützi
MSc ETH Mechanical Engineering

gnuetzi@gmail.com
+41 (0) 77 424 86 45

acki-m commented:
The lib folder should stay!
That's how it works under windows, that is common behavior to put the libs under `lib`
Here how it is done with Qt:
![image](https://user-images.githubusercontent.com/11368160/35356011-7cd48062-014f-11e8-94d8-e90be760a3db.png)

PS: the debug DLL has still no `_d` postfix
Do you have not a virtual machine with windows for testing?

acki-m commented:
This will fix it:
```
    set_target_properties(rttr_core PROPERTIES
                          VERSION ${RTTR_VERSION} SOVERSION ${RTTR_VERSION}
                          DEBUG_POSTFIX _d
                          EXPORT_NAME Core)
```

gabyx commented:
Ehm, the lib folder should have stayed since I did not change anything to the output folders, did you compile and everything went into the lib folder and the binaries into the bin?
I dont have a virtual machine at the moment. I have windows machine, but its not setup for developing, I could check at work tomorrow whats going on...

gabyx commented:
think I found the error, fixing it

acki-m commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+.vscode
+build
```
add new line

acki-m commented on CMake/config.cmake:
```diff
@@ -40,29 +40,59 @@ set(README_FILE "${CMAKE_SOURCE_DIR}/README.md")
 endif()
 
-set(CMAKE_DEBUG_POSTFIX "_d")
+set(CMAKE_DEBUG_POSTFIX CACHE STRING "_d")
```
I think this can be removed.
You have to add the `DEBUG_POSTFIX _d` to all target properties

acki-m commented:
It seems to work now, I wait for the MSVC builds to finish, then I will merge
PS: Why did you changed the ` -stdlib=libstdc++ -static -lstdc++`

gabyx commented:
jup, for me to, on mac os, installation folder looks clean. tests passed...

gabyx commented:
Oh, it dont know why I changed, -stdlib=libstdc++ -static -lstdc++
@Warchant I think commented on this beeing more standart? maybe we should check again... do you know?

acki-m commented:
@gabyx 
I want to squash your 21 commits into one for merge, can you give me a shot summary, what you have done here.

gabyx commented:
Jeah, thats a good thing to do, 
Shot summary: 
- installation folder consistency Linux/Mac/Windows solved
- RPath issues solved for Linux/Mac
- cmake version upgraded to 3.0
I think thats it.

Thanks for your time!
